gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollFailureHandler.java

```diff
@@ -0,0 +1,53 @@
+package io.gravitee.am.gateway.handler.root.resources.endpoint.mfa;
+
+import io.gravitee.am.common.oidc.Parameters;
+import io.gravitee.am.common.utils.ConstantKeys;
+import io.gravitee.am.common.web.UriBuilder;
+import io.gravitee.am.gateway.handler.common.utils.StaticEnvironmentProvider;
+import io.gravitee.am.gateway.handler.common.vertx.utils.UriBuilderRequest;
+import io.gravitee.am.gateway.handler.root.RootProvider;
+import io.gravitee.am.gateway.handler.root.resources.handler.error.AbstractErrorHandler;
+import io.gravitee.am.service.utils.vertx.RequestUtils;
+import io.vertx.rxjava3.core.MultiMap;
+import io.vertx.rxjava3.ext.web.RoutingContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static io.gravitee.am.common.utils.ConstantKeys.MFA_ENROLL_VALIDATION_FAILED;
+import static io.gravitee.am.gateway.handler.common.vertx.utils.UriBuilderRequest.CONTEXT_PATH;
+
+public class MFAEnrollFailureHandler extends AbstractErrorHandler {
+    private static final String ERROR_CODE_VALUE = "enrollment_channel_invalid";
+    public MFAEnrollFailureHandler() {
+        super(RootProvider.PATH_MFA_ENROLL);
+    }
+
+    @Override
+    public void doHandle(RoutingContext routingContext) {
+        handleException(routingContext, routingContext.failure());
```

Comment:
**question**: is there anything we have to remove from the session to have a fresh enrollment state ?

Reply:
I dont think so. Any session variables are set if validation passes